buildSrc/src/main/kotlin/IdeCompatibilityDsl.kt.as41

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+import org.gradle.api.Project
+
+interface CompatibilityPredicate {
+    fun matches(ide: Ide): Boolean
+
+    operator fun invoke(): Boolean = matches(IdeVersionConfigurator.currentIde)
+
+    operator fun invoke(block: () -> Unit): Unit {
+        if (matches(IdeVersionConfigurator.currentIde)) {
+            block()
+        }
+    }
+}
+
+val CompatibilityPredicate.not: CompatibilityPredicate get() = object : CompatibilityPredicate {
+    override fun matches(ide: Ide) = !this@not.matches(ide)
+}
+
+fun CompatibilityPredicate.or(other: CompatibilityPredicate): CompatibilityPredicate = object : CompatibilityPredicate {
+    override fun matches(ide: Ide) = this@or.matches(ide) || other.matches(ide)
+}
+
+enum class Platform : CompatibilityPredicate {
+    P183, P191, P192, P193, P201;
+
+    val version: Int = name.drop(1).toInt()
+
+    override fun matches(ide: Ide) = ide.platform == this
+
+    companion object {
+        operator fun get(version: Int): Platform {
+            return Platform.values().firstOrNull { it.version == version }
+                ?: error("Can't find platform $version")
+        }
+    }
+}
+
+enum class Ide(val platform: Platform) : CompatibilityPredicate {
+    IJ183(Platform.P183),
+    IJ191(Platform.P191),
+    IJ192(Platform.P192),
+    IJ193(Platform.P193),
+    IJ201(Platform.P201),
+
+    AS34(Platform.P183),
+    AS35(Platform.P183),
+    AS36(Platform.P192),
+    AS40(Platform.P193),
+    AS41(Platform.P201);
```

Comment:
We usually prefer to avoid creating bunches for build files to avoid full rebuild after switch operation. By adding additional entry to this file, it's possible to create build files with all supported platforms with conditions in code.